### PR TITLE
Start sending additional memory information per process 

### PR DESCRIPTION
**Pull request by vrdmr**
## Description
Adding proc_statm check to validate the memory values we get from CGroup.

Sample event would look like
```
{
    "SchemaVersion": 1.0,
    "PerfMetrics": {
        "walinuxagent.service": {
            "cpu": {
                "cur_cpu": [
                    0.5708333333333334, 0.531, 0.648, 0.5625, 6, "2019-12-11 00:05:03.961587", "2019-12-11 00:30:05.881054"
                ]
            },
            "memory": {
                "cur_mem": [
                    42011306.666666664, 41652224, 43409408, 41750528.0, 6, "2019-12-11 00:05:03.961690", "2019-12-11 00:30:05.881153"
                ],
                "max_mem": [
                    338530304.0, 338530304, 338530304, 338530304.0, 6, "2019-12-11 00:05:03.961728", "2019-12-11 00:30:05.881193"
                ]
            },
            "proc_statm_memory": {
                "1350 | python3 | /usr/bin/python3 -u /usr/sbin/waagent -daemon": [
                    23064576.0, 23064576, 23064576, 23064576.0, 6, "2019-12-11 00:05:03.961908", "2019-12-11 00:30:05.881369"
                ],
                "1511 | python3 | python3 -u /usr/sbin/waagent -run-exthandlers": [
                    27376981.333333332, 27312128, 27668480, 27318272.0, 6, "2019-12-11 00:05:03.962010", "2019-12-11 00:30:05.881471"
                ]
}}}}
```

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1729)

### CODE REVIEW

File: tests/common/test_cgroups.py [hunk 1]
```diff
@@ -227,8 +257,12 @@ def test_get_metrics(self):
...
+            test_mem_cg.get_memory_usage()
 
-        max_memory_usage = test_mem_cg.get_max_memory_usage()
-        self.assertEqual(0, max_memory_usage)
+        self.assertEqual(e.exception.errno, errno.ENOENT)
+
+        with self.assertRaises(IOError) as e:
+            test_mem_cg.get_max_memory_usage()
+
+        self.assertEqual(e.exception.errno, errno.ENOENT)
```
Comment (vrdmr) [inline tests/common/test_cgroups.py]: Testing the change of behavior - Exception being raised when we don't find cgroup path.

File: tests/ga/test_monitor.py [hunk 2]
```diff
@@ -1120,16 +1127,18 @@ def test_report_event_metrics_sent_for_actual_cgroup(self, patch_report_event, p
         telemetry_event_list = patch_report_event.call_args_list[0][0][0]
 
         for e in telemetry_event_list.events:
+            print([(i.name, i.value) for i in e.parameters])
             details_of_event = [x for x in e.parameters if x.name in
                                 ["Category", "Counter", "Instance", "Value"]]
 
+            # print("details_of_event: ", [(i.name, i.value) for i in details_of_event])
```
Comment (vrdmr) [inline tests/ga/test_monitor.py]: Stray - Will remove it.

File: tests/ga/test_monitor.py [hunk 3]
```diff
@@ -1088,30 +1096,29 @@ def test_report_event_metrics_sent_for_actual_cgroup(self, patch_report_event, p
...
-            BASE_CGROUPS, "cpu", "walinuxagent.extensions", "foobar_1.0.0")))
-        self.assertTrue(CGroupsTelemetry.is_tracked(os.path.join(
-            BASE_CGROUPS, "memory", "walinuxagent.extensions", "foobar_1.0.0")))
+        self.assertTrue(CGroupsTelemetry.is_tracked(os.path.join(BASE_CGROUPS, "cpu", "walinuxagent.extensions", extn_folder_name)))
+        self.assertTrue(CGroupsTelemetry.is_tracked(os.path.join(BASE_CGROUPS, "memory", "walinuxagent.extensions", extn_folder_name)))
 
         for i in range(max_num_polls):
             metrics = CGroupsTelemetry.poll_all_tracked()
-            self.assertEqual(len(metrics), 3)
+            self.assertEqual(len(metrics), 5)
```
Comment (vrdmr) [inline tests/ga/test_monitor.py]: Need to add the explanation of "5"

File: tests/common/test_cgroupstelemetry.py [hunk 4]
```diff
@@ -146,145 +245,131 @@ def test_telemetry_polling_with_active_cgroups(self, *args):
...
-                            self.assertFalse(CGroupsTelemetry.is_tracked("dummy_memory_path_{0}".format(i)))
-
-                        self.assertEqual(CGroupsTelemetry._cgroup_metrics.__len__(), num_extensions)
-                        self._assert_cgroup_metrics_equal(
-                            cpu_usage=[current_cpu] * data_count,
-                            memory_usage=[current_memory] * data_count,
-                            max_memory_usage=[current_max_memory] * data_count)
-                        self.assertEqual(len(metrics), num_extensions * 3)
-                        self._assert_cgroup_polling_metrics_equal(metrics, current_cpu, current_memory, current_max_memory)
+    @patch("azurelinuxagent.common.osutil.default.DefaultOSUtil.get_total_cpu_ticks_since_boot")
```
Comment (narrieta) [inline tests/common/test_cgroupstelemetry.py]: why is this mock needed?

Comment (vrdmr) [reply tests/common/test_cgroupstelemetry.py]: Stray mock left, during solving the merge conflicts while merging `develop`. Thanks. Fixed.

File: azurelinuxagent/ga/monitor.py [hunk 5]
```diff
@@ -288,7 +287,7 @@ def daemon(self):
             self.reset_loggers()
             time.sleep(min_delta)
 
-    def reset_loggers(self):
+    def reset_loggers(self, passed_logger=None):
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply azurelinuxagent/ga/monitor.py]: Done. Will create a new PR for this change.

File: azurelinuxagent/common/exception.py [hunk 6]
```diff
@@ -44,14 +45,18 @@ def __init__(self, msg=None, inner=None):
...
+    When network is not available.
     """
 
     def __init__(self, msg=None, inner=None):
         super(AgentNetworkError, self).__init__(msg, inner)
 
 
 class CGroupsException(AgentError):
+    """
+    When cgroups are not available or could not fetch data from CGroup.
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply azurelinuxagent/common/exception.py]: Generalized it.

File: azurelinuxagent/common/resourceusage.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        python--targettest_resourceusage.py
+        root@vm:/# cat /proc/1392/cmdline | tr "\0" " "
+        python --target test_resourceusage.py
+
+        :return: command line passed to the process string.
+        """
+        cmdline_file_name = PROC_CMDLINE_FILENAME_FORMAT.format(process_id)
+        try:
+            pid_cmdline = fileutil.read_file(cmdline_file_name).replace("\0", " ").strip()
+            pid_cmdline_str = str(pid_cmdline)
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply azurelinuxagent/common/resourceusage.py]: Yes. No need to explicitly cast the pid_cmdline. Removed.

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 8]
```diff
@@ -62,6 +80,14 @@ def _process_cgroup_metric(cgroup_metrics):
             else:
                 processed_extension["memory"] = {"max_mem": CGroupsTelemetry._get_metrics_list(max_memory_usage)}
 
+        if memory_usage_per_process:
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: The `get_memory_metrics`, `get_max_memory_metrics`, and `get_cpu_metrics` return object of Metric(), whereas `get_proc_statm_memory_metrics` return a list of (pid_name_cmdline, metric). If there are no processes tracked, the return value would be an empty list.

Thus, `if memory_usage_per_process` checks if the list returned is empty or not. In the above cases, where I do `foo.count() > 0`, we don't have access to the _data list, and thus I check for count of data-entries.

File: azurelinuxagent/common/cgroup.py [hunk 9]
```diff
@@ -111,6 +111,29 @@ def is_active(self):
 
         return False
 
+    def get_tracked_processes(self):
+        """
+
+        :return: List of Str (Pids)
```
Comment (narrieta) [inline azurelinuxagent/common/cgroup.py]: ":return: List of Str (Pids) or None, if an error occurs"

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 10]
```diff
@@ -151,9 +186,26 @@ def poll_all_tracked():
                         max_memory_usage = cgroup.get_max_memory_usage()
                         CGroupsTelemetry._cgroup_metrics[cgroup.name].add_max_memory_usage(max_memory_usage)
                         metrics.append(MetricValue("Memory", "Max Memory Usage", cgroup.name, max_memory_usage))
+
+                        pids = cgroup.get_tracked_processes()
+
+                        if pids:
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: if pids is not None:

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 11]
```diff
@@ -62,6 +90,14 @@ def _process_cgroup_metric(cgroup_metrics):
             else:
                 processed_extension["memory"] = {"max_mem": CGroupsTelemetry._get_metrics_list(max_memory_usage)}
 
+        if len(memory_usage_per_process) > 0:
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupstelemetry.py]: if not needed (for below will take care of empty)

File: azurelinuxagent/common/resourceusage.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            pid_comm_str = str(pid_comm)
+        except Exception as e:
+            raise ProcessInfoException("Could not get contents from {0}".format(comm_file_name), e)
+
+        return pid_comm_str
+
+
+class ProcessInfoException(AgentError):
+    """
+    When we
```
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/resourceusage.py]: still needs fixing

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 11 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: neat. Didnt think of it earlier. Thanks.

File: azurelinuxagent/common/cgroupstelemetry.py (hunk 10 above)
Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: Made the return streamlined.

File: azurelinuxagent/common/cgroupstelemetry.py [hunk 13]
```diff
@@ -34,16 +52,37 @@ class CGroupsTelemetry(object):
...
+            process_cmdline = ProcessInfo.get_proc_cmdline(process_id) if not None else DEFAULT_PROCESS_COMMANDLINE
+        except Exception as e:
+            logger.periodic_info(EVERY_SIX_HOURS, "[PERIODIC] {0}", ustr(e))
+
+        try:
+            process_name = ProcessInfo.get_proc_name(process_id) if not None else DEFAULT_PROCESS_NAME
+        except Exception as e:
+            logger.periodic_info(EVERY_SIX_HOURS, "[PERIODIC] {0}", ustr(e))
+
+        return process_id + DELIM + process_name + DELIM + process_cmdline
```
Comment (larohra) [inline azurelinuxagent/common/cgroupstelemetry.py]: Small comment, why not use "{0}".format() instead? That makes the string look cleaner and easy to debug

Comment (vrdmr) [reply azurelinuxagent/common/cgroupstelemetry.py]: Though it was simple enough and wanted to keep the structure explicit. I actually felt this was a little unclear to what the generated structure would be.
```
summarized_info = "{1} {0} {2} {0} {3}".format(DELIM, process_id, process_name, process_cmdline)
```

I can fix this in the later PRs if this is not clear enough.

**Pull request merged**